Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/preprocessor/CqlPreprocessor.kt

```diff
@@ -0,0 +1,316 @@
+    }
+
+    override fun visitExpressionDefinition(ctx: ExpressionDefinitionContext): Any {
+        val expressionDefinition = ExpressionDefinitionInfo()
+        expressionDefinition.name = parseString(ctx.identifier())
+        expressionDefinition.context = currentContext
+        expressionDefinition.definition = ctx
+        processHeader(ctx, expressionDefinition)
+        libraryInfo.addExpressionDefinition(expressionDefinition)
+        return expressionDefinition
+    }
+
+    override fun visitFunctionDefinition(ctx: FunctionDefinitionContext): Any {
+        val functionDefinition = FunctionDefinitionInfo()
+        functionDefinition.name = parseString(ctx.identifierOrFunctionIdentifier())
+        functionDefinition.context = currentContext
+        functionDefinition.definition = ctx
+        processHeader(ctx, functionDefinition)
+        libraryInfo.addFunctionDefinition(functionDefinition)
+        return functionDefinition
+    }
+
+    override fun visitNamedTypeSpecifier(ctx: NamedTypeSpecifierContext): NamedTypeSpecifier {
+        val qualifiers = parseQualifiers(ctx)
+        val modelIdentifier: String? = getModelIdentifier(qualifiers)
+        val identifier: String? =
+            getTypeIdentifier(qualifiers, parseString(ctx.referentialOrTypeNameIdentifier()))
+        val typeSpecifierKey = String.format("%s:%s", modelIdentifier, identifier)
+        val resultType = libraryBuilder.resolveTypeName(modelIdentifier, identifier)
+        if (null == resultType) {
+            libraryBuilder.addNamedTypeSpecifierResult(
+                typeSpecifierKey,
+                ResultWithPossibleError.withError()
+            )
+            throw CqlCompilerException(
+                String.format(
+                    "Could not find type for model: %s and name: %s",
+                    modelIdentifier,
+                    identifier
+                )
+            )
+        }
+        val result =
+            of.createNamedTypeSpecifier().withName(libraryBuilder.dataTypeToQName(resultType))
+
+        // Fluent API would be nice here, but resultType isn't part of the model so...
+        result.resultType = resultType
+        libraryBuilder.addNamedTypeSpecifierResult(
+            typeSpecifierKey,
+            ResultWithPossibleError.withTypeSpecifier(result)
+        )
+        return result
+    }
+
+    override fun visitTerminal(node: TerminalNode): Any {
+        var text = node.text
+        val tokenType = node.symbol.type
+        if (cqlLexer.STRING == tokenType || cqlLexer.QUOTEDIDENTIFIER == tokenType) {
+            // chop off leading and trailing ' or "
+            text = text.substring(1, text.length - 1)
+        }
+        return text
+    }
+
+    override fun visitQualifiedIdentifier(ctx: QualifiedIdentifierContext): Any {
+        // Return the list of qualified identifiers for resolution by the containing element
+        val identifiers: MutableList<String?> = ArrayList()
```

Comment:
issue: Some of the type inferences here are incorrect. The ids in this list aren't nullable.

Reply:
Changed this to `val identifiers = ArrayList<String>()`